examples/chat/module/rust/src/lib.rs

```diff
@@ -14,137 +14,92 @@
+            grpc_request_label, grpc_response_label
+        );
+        if grpc_request_label.can_flow_to(&grpc_response_label) {
+            // Retrieve a reference to a room for the request label, or create one if it does not
+            // already exist.
+            let room_node_sender = self
+                .rooms
+                .entry(grpc_request_label.clone())
+                .or_insert_with(|| create_room_node(&grpc_request_label));
+            // Forward the incoming gRPC invocation to the appropriate room node.
+            room_node_sender.send(&grpc_invocation)
+        } else {
+            // If the caller is trying to send a message with a label such that it would not be able
+            // to receive it itself, do not even attempt to process the request and instead return
+            // an error immediately.
+            warn!("gRPC request and response have inconsistent labels");
+            let mut grpc_response_writer =
+                oak::grpc::ChannelResponseWriter::new(grpc_invocation.response_sender);
+            grpc_response_writer.close(Err(oak::grpc::build_status(
+                oak::grpc::Code::PERMISSION_DENIED,
+                "inconsistent labels",
+            )))
         }
     }
+}
 
-    fn subscribe(&mut self, req: SubscribeRequest, mut writer: grpc::ChannelResponseWriter) {
-        info!("subscribing to room");
-        match self.rooms.get(&req.room_id) {
-            None => {
-                writer
-                    .close(room_id_not_found_err())
-                    .expect("could not close channel");
-            }
-            Some(room) => {
-                info!("new subscription to room {:?}", req.room_id);
-                let command = Command::Join(writer.handle());
-                room.sender
-                    .send(&command)
-                    .expect("could not send command to room Node");
-            }
-        };
-    }
-
-    fn send_message(&mut self, req: SendMessageRequest) -> grpc::Result<Empty> {
-        info!("sending message to room");
-        match self.rooms.get(&req.room_id) {
-            None => room_id_not_found_err(),
-            Some(room) => {
-                info!("new message to room {:?}", req.room_id);
-                let message_bytes = req
-                    .get_message()
-                    .write_to_bytes()
-                    .expect("could not convert message to bytes");
-                let command = Command::SendMessage(message_bytes);
-                room.sender
-                    .send(&command)
-                    .expect("could not send command to room Node");
-                Ok(Empty::new())
-            }
-        }
-    }
+fn create_room_node(grpc_invocation_label: &oak::label::Label) -> oak::io::Sender<Invocation> {
+    let (wh, rh) = oak::channel_create_with_label(grpc_invocation_label).unwrap();
+    oak::node_create_with_label("app", "oak_main", rh, grpc_invocation_label)
```

Comment:
Are labels intended to be static (specified up-front at channel/node creation time) or dynamic (extended when labelled data arrives) or both?

I guess I'd imagined more dynamic behaviour, which seems like it would make the explicit specification of labels here unnecessary &ndash; the new per-Room Node would start with an empty label, but would acquire the room label on its first read of a request message.

(Aside: it's also quite tricky to keep track of the different levels of indirection: the channel that the `Invocation` arrives on has no labels, but the `Invocation` itself contains two channels which do have the labels, and this `Invocation` is forwarded down another channel &ndash; which again can stay label-free because the labelled data is in an inner channel.  Not sure what to do about that, though, other than having clear docs.)

Reply:
I am trying to avoid "floating" labels that are automatically added when reading from channels, since that may introduce [covert channels](https://www.cs.tau.ac.il/~tromer/papers/flumecsp.pdf#page=5).

That said, it would be possible to implement this (and in fact it was one of my earlier attempts) so that a room node is created with "public" label, and then explicitly changes its label based on the first message it receives. I was not a big fan of it though, because I could not find a corresponding mechanism for channels, so I still had to let the channel creator specify the label for the channel upfront, and that should definitely not change once set. Anyways, this is not set in stone yet, happy to revisit and try out different models if this is confusing.

A more drastic approach could be to get rid of channels entirely, and give handles to nodes, effectively melding channels and nodes together, but that may have other side-effects (e.g. client streaming, as you mentioned a while ago).